src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/IncludeRewriter.cs

```diff
@@ -55,12 +59,164 @@ public override Expression VisitSqlRoot(SqlRootExpression expression, object con
                 }
             }
 
-            if (containsInclude)
+            reorderedExpressions.Add(IncludeUnionAllExpression);
+            return new SqlRootExpression(reorderedExpressions, expression.DenormalizedExpressions);
+        }
+
+        private class IncludeExpressionTopologicalSort
+        {
+            // Based on Khan's algorithm. See https://en.wikipedia.org/wiki/Topological_sorting.
+            // The search queries are acyclic.
```

Comment:
How do we know that they are acyclic?

Reply:
Perhaps a better way to say it was that we are not supporting cyclic queries (cyclic queries are actually recursive queries). According to the Fhir Team, it's up to the server implementor to decide how deep.
We agreed that recursive queries are out of scope. 

If it is a cyclic query (recursive query), we can throw an error. It's really a product decision.

See the FHIR chat: https://chat.fhir.org/#narrow/stream/179166-implementers/topic/_include.3Arecurse.20(_include.3Aiterate).20ambiguity

Reply:
Yes, I think we should throw an error if we encounter a cycle.